src/tools/run-make-support/src/diff/mod.rs

```diff
@@ -0,0 +1,64 @@
+use similar::TextDiff;
+use std::path::Path;
+
+#[cfg(test)]
+mod tests;
+
+pub fn diff() -> Diff {
+    Diff::new()
+}
+
+#[derive(Debug)]
+pub struct Diff {
+    expected: Option<String>,
+    actual: Option<String>,
+}
+
+impl Diff {
+    /// Construct a bare `diff` invocation.
+    pub fn new() -> Self {
+        Self { expected: None, actual: None }
+    }
+
+    /// Specify the expected output for the diff from a file.
+    pub fn expected_file<P: AsRef<Path>>(&mut self, path: P) -> &mut Self {
+        let content = std::fs::read_to_string(path).expect("failed to read file");
+        self.expected = Some(content);
+        self
+    }
+
+    /// Specify the expected output for the diff from a given text string.
+    pub fn expected_text<T: AsRef<[u8]>>(&mut self, text: T) -> &mut Self {
+        self.expected = Some(String::from_utf8_lossy(text.as_ref()).to_string());
+        self
+    }
+
+    /// Specify the actual output for the diff from a file.
+    pub fn actual_file<P: AsRef<Path>>(&mut self, path: P) -> &mut Self {
+        let content = std::fs::read_to_string(path).expect("failed to read file");
+        self.actual = Some(content);
+        self
+    }
+
+    /// Specify the actual output for the diff from a given text string.
+    pub fn actual_text<T: AsRef<[u8]>>(&mut self, text: T) -> &mut Self {
+        self.actual = Some(String::from_utf8_lossy(text.as_ref()).to_string());
+        self
+    }
+
+    /// Executes the diff process, prints any differences to the standard error.
+    pub fn run(&self) {
+        let output = self.do_run();
+        if !output.is_empty() {
+            panic!("{}", output)
```

Comment:
I'm not sure if this should be replaced with `handle_failed_output()` to print a uniform error format.

Reply:
I think we might want to give a bit more context, as in something like

```text
test failed: $EXPECTED is different from $ACTUAL

--- $EXPECTED
+++ $ACTUAL
@@ -1,3 +1,3 @@
 foo
+baz
 bar
-baz
```

`handle_failed_output` is more for dumping a failed `Command` (its entire invocation, its stderr/stdout and exit status).